---
title: "Meta-anlysis overview"
author: "Daoming Wang"
date: "2023/12/04"
output:
  html_document: 
    theme: flatly
    highlight: espresso
    toc: true
    toc_depth: 4
    toc_float: true
  word_document: default
  pdf_document:
    includes:
      in_header: header.tex
      keep_tex: yes
      latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## 1 Preparation

### 1.1 Library

```{r 1.1}
suppressMessages(source("functions.R"))
```

### 1.2 Input

```{r 1.2}
mwasMeta <- fread("00.rawData/mwas/metaAnalysis/all.EMMAX.mwas.meta",sep = "\t")
mwasSumm.long <- read.table("00.rawData/mwas/metaAnalysis/all.EMMAX.summ.tsv", sep = "\t", header = T, quote = "")
mwasMeta.cl.pt.full.filtered <-read.table("08.metaAnalysisQC/mwasMeta.cl.pt.full.filtered.tsv",sep = "\t", header = T, check.names = F)
gwasPara  <- read.table("06.MWAS_preparation/mwasPara.tsv", sep = "\t", header = F)
phenInfo<-read.table("02.phenotype_summary/phenInfoSumm.tsv", sep = "\t",header = T, check.names = F)
speInfo<-read.table("03.genotypeSummary/speInfoSumm.tsv",sep = "\t", header = T, check.names = F)

load("01.cleanData/phenotype/full_phen.RData")
sampleCohorts<-read.table("00.rawData/cohortList/sampleCohorts.tsv",header = F)
snpAnno <- fread("00.rawData/snpAnno/dmp.lld1.lld2.300ob.500fg.ibd.tzfg.base1.perCohort.geno0.9.maf0.05.snpN1000.annotation.tsv", sep = "\t", header = F)
snpInfo <- fread("00.rawData/snpAnno/dmp.lld1.lld2.300ob.500fg.ibd.tzfg.base1.perCohort.geno0.9.maf0.05.snpN1000.pvar.tsv")
geneAnno.df <- readRDS("01.cleanData/annotation/geneAnno.df.rds")
```

### 1.3 Setting

```{r 1.3}
snpAnno$Species <- snpAnno$V1 %>% str_replace_all("_.*", "")
snpAnno$Impact  <- snpAnno$V3 %>% match(annotationType) %>% annotationImpact[.]
snpAnno$ImpactLevel <- snpAnno$V3 %>% match(annotationType) %>% annotationLevel[.]
geneAnno <- snpAnno[!duplicated(snpAnno$V2),c(5,2,4)]
colnames(geneAnno) <- c("Species", "Gene", "Annotation")

# Annotation proportion 
snpAnnoSumm.long <- table(snpAnno$Species, snpAnno$Impact) %>% as.data.frame()
snpAnnoSumm <- snpAnnoSumm.long %>% pivot_wider(names_from = Var2, values_from = Freq)
snpAnnoSumm$TotalN <- snpAnnoSumm$High + snpAnnoSumm$Moderate + snpAnnoSumm$Low
snpAnnoSumm$HighProp <- snpAnnoSumm$High/snpAnnoSumm$TotalN
snpAnnoSumm$ModerateProp <- snpAnnoSumm$Moderate/snpAnnoSumm$TotalN
snpAnnoSumm$LowProp <- snpAnnoSumm$Low/snpAnnoSumm$TotalN
snpAnnoSumm$HighModerateProp <- snpAnnoSumm$HighProp + snpAnnoSumm$ModerateProp
```

## 2 Clean table

```{r 2}
# Clean meta-analysis associations
mwasMeta.cl <- mwasMeta
colnames(mwasMeta.cl)[1]<-"Phenotype"
mwasMeta.cl<-mwasMeta.cl[,-3]
mwasMeta.cl$Algorithm <- "REML"
mwasMeta.cl$Phenotype <- mwasMeta.cl$Phenotype %>% str_replace_all(".EMMAX", "")

mwasMeta.cl$CHR <- mwasMeta.cl$SNP %>% str_replace_all("_.*", "")
mwasMeta.cl$BP  <- mwasMeta.cl$SNP %>% str_replace_all(".*_", "")
mwasMeta.cl<-mwasMeta.cl[-grep("TLCA|TUDCA|TLCA_p|TUDCA_p", mwasMeta.cl$Phenotype),]

mwasMeta.cl$A1 <- mwasMeta.cl$SNP %>% match(snpInfo$V3) %>% snpInfo$V4[.]
mwasMeta.cl$A2 <- mwasMeta.cl$SNP %>% match(snpInfo$V3) %>% snpInfo$V5[.]

mwasMeta.cl$Allele1<-mwasMeta.cl$Allele1 %>% toupper()
mwasMeta.cl$Allele2<-mwasMeta.cl$Allele2 %>% toupper()

# tt <- mwasMeta.cl[,-c(17:57)] %>% na.omit()
mwasMeta.cl.sign <- ifelse(mwasMeta.cl$A1 == mwasMeta.cl$Allele1, 1, -1)
mwasMeta.cl$Effect <- mwasMeta.cl$Effect * mwasMeta.cl.sign

# Clean summary
mwasSumm <- pivot_wider(mwasSumm.long, names_from = Parameter, values_from = Value) %>% as.data.frame %>% dplyr::select(1:13)
for (i in c(2:13)) {
  mwasSumm[,i] <- as.numeric(mwasSumm[,i])
}

mwasSumm<-mwasSumm[-grep("TLCA|TUDCA|TLCA_p|TUDCA_p", mwasSumm$Phenotype),]
mwasSumm$Phenotype <- mwasSumm$Phenotype %>% str_replace_all(".EMMAX", "")


mwasSumm$Lowest.P.meta[is.infinite(mwasSumm$Lowest.P.meta)]<-NA
mwasSumm$Lowest.P.meta.nohet[is.infinite(mwasSumm$Lowest.P.meta.nohet)]<-NA

for (i in 1:nrow(mwasSumm)) {
#  i <- 1
  mwasMeta.cl.pt.full.filtered_i <- mwasMeta.cl.pt.full.filtered %>% dplyr::filter(Phenotype== mwasSumm$Phenotype[i])
  mwasMeta.cl.pt.full.filtered_nohet_i <- mwasMeta.cl.pt.full.filtered_i %>% dplyr::filter(HetPVal>0.05) 

  mwasSumm$Lowest.P.meta[i]       <- min(mwasMeta.cl.pt.full.filtered_i$P.value, na.rm = T)
  mwasSumm$N.SNP.metaStudyWideSig[i] <- sum(mwasMeta.cl.pt.full.filtered_i$P.value < studyWideP, na.rm = T)
  mwasSumm$N.SNP.metaMetaWideSig[i]  <- sum(mwasMeta.cl.pt.full.filtered_i$P.value < metaWideP &
                                              mwasMeta.cl.pt.full.filtered_i$P.value > studyWideP, na.rm = T)
  mwasSumm$N.SNP.metaSpeWideSig[i]   <- sum(mwasMeta.cl.pt.full.filtered_i$P.value < speWideP  &
                                              mwasMeta.cl.pt.full.filtered_i$P.value > metaWideP, na.rm = T)
  
  mwasSumm$Lowest.P.meta.nohet[i]       <- min(mwasMeta.cl.pt.full.filtered_nohet_i$P.value, na.rm = T)
  mwasSumm$N.SNP.metaStudyWideSig.nohet[i] <- sum(mwasMeta.cl.pt.full.filtered_nohet_i$P.value < studyWideP, na.rm = T)
  mwasSumm$N.SNP.metaMetaWideSig.nohet[i]  <- sum(mwasMeta.cl.pt.full.filtered_nohet_i$P.value < metaWideP &
                                                    mwasMeta.cl.pt.full.filtered_nohet_i$P.value > studyWideP, na.rm = T)
  mwasSumm$N.SNP.metaSpeWideSig.nohet[i]   <- sum(mwasMeta.cl.pt.full.filtered_nohet_i$P.value < speWideP &
                                                    mwasMeta.cl.pt.full.filtered_nohet_i$P.value > metaWideP, na.rm = T)
}

mwasSumm$Lowest.P.meta[is.infinite(mwasSumm$Lowest.P.meta)] <- NA
mwasSumm$Lowest.P.meta.nohet[is.infinite(mwasSumm$Lowest.P.meta.nohet)] <- NA

mwasSumm$N.SNP.metaSig <- mwasSumm$N.SNP.metaSpeWideSig + mwasSumm$N.SNP.metaMetaWideSig + mwasSumm$N.SNP.metaStudyWideSig 
mwasSumm$N.SNP.metaSig.nohet <- mwasSumm$N.SNP.metaSpeWideSig.nohet + mwasSumm$N.SNP.metaMetaWideSig.nohet + mwasSumm$N.SNP.metaStudyWideSig.nohet
mwasSumm <- mwasSumm %>% dplyr::filter(N.SNP.metaSig > 0)

mwasSumm.sort <- dplyr::arrange(mwasSumm, Lowest.P.meta)
mwasSumm.sort$Algorithm <- "REML"
#mwasSumm.sort$N.SNP.Sig <- mwasSumm.sort$N.SNP.metaStudyWideSig+mwasSumm.sort$N.SNP.metaMetaWideSig+mwasSumm.sort$N.SNP.metaSpeWideSig
mwasSumm.sort <- cbind(mwasSumm.sort$Phenotype %>% match(phenInfo$UnifiedName) %>% phenInfo[.,c(1,2,3,5,6,7)], mwasSumm.sort)

if(!dir.exists("09.metaAnalysisOverview/")){dir.create("09.metaAnalysisOverview/")}
write.table(mwasSumm.sort,"09.metaAnalysisOverview/EMMAX.mwasSumm.tsv", sep = "\t", row.names = F, col.names = T, quote = F)

mwasMeta.cl.pt.full.filtered <- cbind(mwasMeta.cl.pt.full.filtered, 
                                      mwasMeta.cl.pt.full.filtered$Gene %>% match(geneAnno.df$ID) %>% geneAnno.df[.,c(2,4,6:8,10)])

mwasMeta.cl.pt.full.filtered$MetaSigLevel <- "Species-wide"
mwasMeta.cl.pt.full.filtered$MetaSigLevel[mwasMeta.cl.pt.full.filtered$P.value < metaWideP] <- "Metagenome-wide"
mwasMeta.cl.pt.full.filtered$MetaSigLevel[mwasMeta.cl.pt.full.filtered$P.value < studyWideP] <- "Study-wide"

if(!dir.exists("09.metaAnalysisOverview")){dir.create("09.metaAnalysisOverview")}
write.table(mwasMeta.cl.pt.full.filtered,"09.metaAnalysisOverview/mwasMeta.cl.pt.full.filtered.tsv", sep = "\t", row.names = F, col.names = T, quote = F)
```

## 3 Replicable association

### 3.1 Association categories

```{r 3.1}
pdf("09.metaAnalysisOverview/mwasMeta.sig.sigLevel.pie.pdf", width = 6, height = 6)
ggpie(data = mwasMeta.cl.pt.full.filtered[,c(93:94,121)], group_key = "MetaSigLevel", count_type = "full",
               fill_color = pal_material("orange")(6)[c(6,4,3)],
               label_info = "all", label_type = "horizon", label_split = NULL,
               label_size = 4, label_pos = "in", label_threshold = 10,
               border_size = 0)
dev.off()

pdf("09.metaAnalysisOverview/mwasMeta.sig.repN.pie.pdf", width = 6, height = 6)
ggpie(data = mwasMeta.cl.pt.full.filtered[,93:94], group_key = "Nominal_sig_cohort_N.QC", count_type = "full",
      fill_color = pal_material("blue")(6)[c(2,3,4,6)],
      label_info = "all", label_type = "horizon", label_split = NULL,
      label_size = 4, label_pos = "in", label_threshold = 10,
      border_size = 0)
dev.off()

pdf("09.metaAnalysisOverview/mwasMeta.sig.Impact.pie.pdf", width = 6, height = 6)
ggpie(data = mwasMeta.cl.pt.full.filtered[,111:114], group_key = "ImpactLevel", count_type = "full",
      fill_color = pal_material("red")(6)[c(2,4,6)],
      label_info = "all", label_type = "horizon", label_split = NULL,
      label_size = 4, label_pos = "in", label_threshold = 10,
      border_size = 0)
dev.off()

pdf("09.metaAnalysisOverview/mwasMeta.sig.FunctionKnown.pie.pdf", width = 6, height = 6)
ggpie(data = mwasMeta.cl.pt.full.filtered[,111:114], group_key = "FunctionKnown", count_type = "full",
      fill_color = pal_material("green")(6)[c(6,2)],
      label_info = "all", label_type = "horizon", label_split = NULL,
      label_size = 4, label_pos = "in", label_threshold = 10,
      border_size = 0)
dev.off()
```

### 3.2 Impact of SNPs

```{r 3.2}
# Summary of significant associations
mwasMeta.cl.pt.summ.long <- table(mwasMeta.cl.pt.full.filtered$Phenotype, mwasMeta.cl.pt.full.filtered$CHR, mwasMeta.cl.pt.full.filtered$ImpactLevel) %>% as.data.frame()
mwasMeta.cl.pt.summ      <- mwasMeta.cl.pt.summ.long %>% pivot_wider(names_from = Var3, values_from = Freq)
mwasMeta.cl.pt.summ$TotalN <- mwasMeta.cl.pt.summ$High + mwasMeta.cl.pt.summ$Moderate + mwasMeta.cl.pt.summ$Low
mwasMeta.cl.pt.summ$HighProp <- mwasMeta.cl.pt.summ$High/mwasMeta.cl.pt.summ$TotalN
mwasMeta.cl.pt.summ$ModerateProp <- mwasMeta.cl.pt.summ$Moderate/mwasMeta.cl.pt.summ$TotalN
mwasMeta.cl.pt.summ$LowProp <- mwasMeta.cl.pt.summ$Low/mwasMeta.cl.pt.summ$TotalN
mwasMeta.cl.pt.summ$HighModerateProp <- mwasMeta.cl.pt.summ$HighProp + mwasMeta.cl.pt.summ$ModerateProp
mwasMeta.cl.pt.summ.filt <- mwasMeta.cl.pt.summ %>% dplyr::filter(TotalN >= 10)
mwasMeta.cl.pt.summ.filt$PhenClass <- mwasMeta.cl.pt.summ.filt$Var1 %>% match(phenInfo$UnifiedName) %>% phenInfo$Class[.]
snpAnno.summ<-table(snpAnno$V3) %>% as.data.frame()
snpAnno.summ$ImpactLevel <- snpAnno.summ$Var1 %>% match(annotationType) %>% annotationLevel[.]


pdf("09.metaAnalysisOverview/All.SNP.annotation.pie.pdf")
ggnestedpie(data = snpAnno, group_key = c("ImpactLevel", "V3"), count_type = "full",
            outer_fill_color = c("#E64B35","#3C5488", "#00A087"), border_size = 0,
            outer_label_type = "horizon", outer_label_info = "all",outer_label_pos = "out",
            inner_label = F)
dev.off()

snpAnnoSumm.filt <- snpAnnoSumm[snpAnnoSumm$Var1 %in% mwasMeta.cl.pt.summ.filt$Var2,]
speAnnoSumm.filt.order <- order(snpAnnoSumm.filt$HighModerateProp, decreasing = T) %>% snpAnnoSumm.filt$Var1[.]

p_prop_high_moder_spe <- ggplot() + 
  geom_bar(data=snpAnnoSumm.filt, aes(Var1, HighModerateProp),stat="identity", color = "#baccd9",fill = "#baccd9") +
  geom_point(data=mwasMeta.cl.pt.summ.filt, aes(Var2, HighModerateProp, size = log10(TotalN), color = PhenClass), alpha = 0.7) +
  scale_x_discrete(limits = speAnnoSumm.filt.order) +
  scale_color_manual(breaks = phenClassOrder,
                     values = phenClassColor)+
  ylab("Propotion of high/moderate impact SNPs")+
  xlab("Species")+
  theme_linedraw()+
  theme(legend.position = "right",
        legend.justification = c(0, 1),
        legend.text = element_text(size = 10),
        plot.subtitle = element_text(vjust = 1), 
        plot.caption = element_text(vjust = 1), 
        axis.line = element_line(colour = 'black'), 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = NA),
        panel.border = element_blank(),
        axis.line.x = element_line(colour = 'white'),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

pdf("09.metaAnalysisOverview/prop_high_moderate_bar.pdf", width = 10, height = 3.5)
print(p_prop_high_moder_spe)
dev.off()
```

### 3.3 Prioritization

```{r 3.3}
mwasMeta.cl.pt.top <- mwasMeta.cl.pt.full.filtered %>% dplyr::filter(Nominal_sig_cohort_N >= 2  & (ImpactLevel != "Low") & (Gene != Annotation) )
write.table(mwasMeta.cl.pt.top, "09.metaAnalysisOverview/EMMAX.mwasMeta.clean.prioritized.tsv",sep = "\t", col.names = T, row.names = F, quote = F)

# Number of prioritized association for each phenotype
mwasMeta.cl.pt.top.N     <- table(mwasMeta.cl.pt.top$Phenotype) %>% as.data.frame()
colnames(mwasMeta.cl.pt.top.N) <- c("Phenotype",  "Count")
mwasMeta.cl.pt.top.N_order <- mwasMeta.cl.pt.top.N$Phenotype[order(mwasMeta.cl.pt.top.N$Count, decreasing = T)]

mwasMeta.cl.pt.top.N.rep <- table(mwasMeta.cl.pt.top$Phenotype, mwasMeta.cl.pt.top$Nominal_sig_cohort_N) %>% as.data.frame()
colnames(mwasMeta.cl.pt.top.N.rep) <- c("Phenotype", "Nominal_sig_cohort_N", "Count")
mwasMeta.cl.pt.top.N.rep$Nominal_sig_cohort_N <- factor(mwasMeta.cl.pt.top.N.rep$Nominal_sig_cohort_N, levels = c(5,4,3,2))


p_mwasMeta.cl.pt.top.N.rep <- mwasMeta.cl.pt.top.N.rep %>% 
  ggplot(aes(Phenotype, Count)) +
  geom_bar(aes(fill = Nominal_sig_cohort_N),color = "white",size = 0, stat="identity")+
  scale_x_discrete(limits = mwasMeta.cl.pt.top.N_order[1:38])+
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(name=NULL, 
                     breaks = c(5, 4,3,2),
                     labels = c(5, 4,3,2),
                     values = pal_material("blue")(6)[c(6,4,3,2)])+
  ylab("Number of associated SNPs")+
  theme_bw()+
  theme(legend.position = "bottom",
        legend.justification = c(0, 1),
        legend.text = element_text(size = 10),
        plot.subtitle = element_text(vjust = 1), 
        plot.caption = element_text(vjust = 1), 
        axis.line = element_line(colour = 'white'), 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = NA),
        panel.border = element_blank(),
        axis.line.x = element_line(colour = 'white'),
        axis.title.x=element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.ticks.x=element_blank())

pdf("09.metaAnalysisOverview/mwasMeta.cl.pt.top.N.rep_bar.pdf", width = 8, height = 3.5)
print(p_mwasMeta.cl.pt.top.N.rep)
dev.off()

# Number of associated genes
mwasMeta.cl.pt.top.gene <- mwasMeta.cl.pt.top[,c(16, 18, 110, 112)] %>% .[!duplicated(.[,c(1,3)]),]
mwasMeta.cl.pt.top.gene$Species_gene <- paste(mwasMeta.cl.pt.top.gene$CHR, mwasMeta.cl.pt.top.gene$Annotation, sep = "__")
mwasMeta.cl.pt.top.gene.count <- table(mwasMeta.cl.pt.top.gene$Gene) %>% sort(decreasing = T) %>% table %>% as.data.frame()
colnames(mwasMeta.cl.pt.top.gene.count) <- c("AssociatedPhenNum", "GeneCount")

p_mwasMeta.cl.pt.top.gene.count<-mwasMeta.cl.pt.top.gene.count[-1,] %>%
  ggplot(aes(AssociatedPhenNum, GeneCount)) +
  geom_bar(fill = "#257CB2", size = 0, stat="identity")+
  scale_y_continuous(expand = c(0,0)) +
  xlab("Associated phenotype number")+
  ylab("Number of gene")+
  theme_bw()+
  theme(legend.position = "bottom",
        legend.justification = c(0, 1),
        legend.text = element_text(size = 10),
        plot.subtitle = element_text(vjust = 1), 
        plot.caption = element_text(vjust = 1), 
        axis.line = element_line(colour = 'white'), 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = NA),
        panel.border = element_blank(),
        axis.line.x = element_line(colour = 'white'),
        axis.text.x = element_text(angle = 0, hjust = 1),
        axis.ticks.x=element_blank())

pdf("09.metaAnalysisOverview/mwasMeta.cl.pt.top.gene.count_bar.pdf", width = 5, height = 2.5)
print(p_mwasMeta.cl.pt.top.gene.count)
dev.off()

mwasMeta.cl.pt.top.gene.count.top <- mwasMeta.cl.pt.top.gene.count <- table(mwasMeta.cl.pt.top.gene$Gene) %>% sort(decreasing = T) 
mwasMeta.cl.pt.top.gene.count.top.df <- data.frame(Gene = names(mwasMeta.cl.pt.top.gene.count.top),
                                                   Count = as.numeric(mwasMeta.cl.pt.top.gene.count.top),
                                                   Annotation = names(mwasMeta.cl.pt.top.gene.count.top) %>% match(geneAnno$Gene) %>% geneAnno$Annotation[.])

mwasMeta.cl.pt.top.anno <- mwasMeta.cl.pt.top.gene %>% .[!duplicated(.[,c(1,3,4)] ), c(1,2,4)]
mwasMeta.cl.pt.top.anno.count <- table(mwasMeta.cl.pt.top.anno$Phenotype, mwasMeta.cl.pt.top.anno$Annotation) %>% as.data.frame()
```

### 3.4 Meta-analysis circos plot

```{r 3.4}
mwasSumm.meta <- mwasSumm.sort[which(mwasSumm.sort$N.SNP.useForMetaAnalysis>0 & mwasSumm.sort$Lowest.P.meta < 6.47E-6),] %>%
  .[order(rowSums(.[,13:15]), decreasing = T), ]
write.table(mwasSumm.meta, "09.metaAnalysisOverview/mwasSumm.meta.tsv", sep = "\t", row.names = F, col.names = T, quote = F)

pdf("09.metaAnalysisOverview/EMMAX.mwasMeta.summary.circos.pdf", width = 10, height = 8)
# Initiate circos plot
circos.par("start.degree" = 180, "gap.after" = 180)
circos.initialize('a', xlim = c(0, length(mwasSumm.meta$Phenotype)))

# Phenotype class
Class_col = c( "Anthropology" = "#257CB2",           "CurrentExposure" = "#E93E3F", "Diet" = "#4F9F3B",
               "EarlyLifeExposure" = "#FE8B15",      "Metabolite" = "#73489F",
               "GeographySocioeconomics" = "#A6CEE3","HealthDisease" = "#F68181",
               "MedicalMeasurement" = "#A5D981",     "MedicationUse" = "#F7F599", 
               "Stool" = "#10aec2")

phen.class <- match(mwasSumm.meta$Phenotype, phenInfo$UnifiedName) %>% phenInfo$Class[.]
circos.track(ylim = c(0, 1),cell.padding = c(0,0,0,0),
             track.height = 0.05, bg.border = NA, bg.col = adjustcolor("white", alpha.f = 0.02), track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = rep(1, length(phen.class))
               circos.barplot(value, 1:length(phen.class), col = match(phen.class, names(Class_col)) %>% Class_col[.], 
                              border = adjustcolor("white", alpha.f = 0),bar_width = 1)
               circos.text(1:nrow(mwasSumm.meta), 1, 
                           paste0(mwasSumm.meta$Phenotype),
                           facing = "clockwise", cex = 0.6,
                           adj = c(0, 0.5), niceFacing = TRUE)
             })

# Number of SNPs that are study-wide significant 
circos.track(ylim = c(0, log10(max(mwasSumm.meta$N.SNP.metaStudyWideSig))),cell.padding = c(0,0,0.02,0),
             track.height = 0.1,bg.border = NA, bg.col = adjustcolor("#ed5a65", alpha.f = 0.05), track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = log10(mwasSumm.meta$N.SNP.metaStudyWideSig)
               value[is.infinite(value)]<-0
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = "#ed5a65", border = adjustcolor("#ed5a65", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F, at = seq(0, 3, by = 1))
               circos.text(105, 1, 
                           paste0("study-wide significant SNP N"),
                           facing = "bending.inside", cex = 0.8,
                           adj = c(1, -0.5), niceFacing = TRUE)
             })

# Number of SNPs that are metagenome-wide significant 
circos.track(ylim = c(0, log10(max(mwasSumm.meta$N.SNP.metaMetaWideSig))),cell.padding = c(0,0,0.02,0),
             track.height = 0.1,bg.border = NA, bg.col = adjustcolor("#fecc11", alpha.f = 0.05),  track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = log10(mwasSumm.meta$N.SNP.metaMetaWideSig)
               value[is.infinite(value)]<-0
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = "#fecc11", border = adjustcolor("#fecc11", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F, at = seq(0, 3, by = 1))
               circos.text(105, 1, 
                           paste0("metagenome-wide significant SNP N"),
                           facing = "bending.inside", cex = 0.8,
                           adj = c(1, -0.5), niceFacing = TRUE)
             })

# Number of SNPs that are species-wide significant 
circos.track(ylim = c(0, log10(max(mwasSumm.meta$N.SNP.metaSpeWideSig))),cell.padding = c(0,0,0.02,0),
             track.height = 0.1,bg.border = NA,bg.col = adjustcolor("#2775b6", alpha.f = 0.05), track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = log10(mwasSumm.meta$N.SNP.metaSpeWideSig)
               value[is.infinite(value)]<-0
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = "#2775b6", border = adjustcolor("#2775b6", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F, at = seq(0, 3, by = 1))
               circos.text(105, 1, 
                           paste0("species-wide significant SNP N"),
                           facing = "bending.inside", cex = 0.8,
                           adj = c(1, -0.5), niceFacing = TRUE)
             })

# Number of SNPs that are significant 
circos.track(ylim = c(0, log10(max(rowSums(mwasSumm.meta[,13:15])))),cell.padding = c(0,0,0.02,0),
             track.height = 0.1,bg.border = NA,bg.col = adjustcolor("black", alpha.f = 0.05), track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = log10(rowSums(mwasSumm.meta[,13:15]))
               value[is.infinite(value)]<-0
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = "black", border = adjustcolor("black", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F)
               circos.text(105, 1, 
                           paste0("significant SNP N"),
                           facing = "bending.inside", cex = 0.8,
                           adj = c(1, -0.5), niceFacing = TRUE)
             })
# Number of SNPs used for meta-analysis
circos.track(ylim = c(0, max(mwasSumm.meta$N.SNP.useForMetaAnalysis)),cell.padding = c(0,0,0.02,0),
             track.height = 0.1,bg.border = NA,bg.col = adjustcolor("#847c74", alpha.f = 0.05), track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = mwasSumm.meta$N.SNP.useForMetaAnalysis
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = "#847c74", border = adjustcolor("#847c74", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F)
               circos.text(105, 1, 
                           paste0("meta-analysis SNP N"),
                           facing = "bending.inside", cex = 0.8,
                           adj = c(1, -0.5), niceFacing = TRUE)
             })

# Number of SNPs
circos.track(ylim = c(0, max(mwasSumm.meta$N.SNP)),cell.padding = c(0,0,0.02,0),
             bg.border = NA,  track.margin = c(0,0),
             panel.fun = function(x, y) {
               value = data.frame(N.SNP.detectedIn1Cohort     = mwasSumm.meta$N.SNP.detectedIn1Cohort,
                                  N.SNP.detectedInMultiCohort = mwasSumm.meta$N.SNP.detectedInMultiCohort ) %>% as.matrix
               circos.barplot(value, 1:length(mwasSumm.meta$Phenotype), col = c("#0f95b0","#ef475d"), border = adjustcolor("white", alpha.f = 0))
               circos.yaxis(side = "left", labels.cex = 0.5,labels.niceFacing = F)
             })

circos.clear()
dev.off()
```

### 3.5 Meta-analysis heatmap

```{r 3.5}
colnames(mwasMeta)[1]<-"Phenotype"
mwasMeta<-mwasMeta[,-3]
mwasMeta$Algorithm <- "REML"
mwasMeta$Phenotype <- mwasMeta$Phenotype %>% str_replace_all(".EMMAX", "")

mwasMeta$SpeciesId <- mwasMeta$SNP %>% str_replace_all("_.*", "")
mwasMeta$CHR <- mwasMeta$SNP %>% str_replace_all("_.*", "")
mwasMeta$Pos <- mwasMeta$SNP %>% str_replace_all(".*_", "")
mwasMeta<-mwasMeta[-grep("TLCA|TUDCA", mwasMeta$Phenotype),]

mwasMeta.sig <- paste(mwasMeta.cl.pt.full.filtered$Phenotype, mwasMeta.cl.pt.full.filtered$CHR) %>%
  match(paste(mwasMeta$Phenotype, mwasMeta$CHR)) %>%
  mwasMeta[.,]
#mwasMeta.sig <- mwasMeta[which(mwasMeta$`P-value`<6.47E-6),]
#mwasMeta.sig <- mwasMeta[which(mwasMeta$`P-value`<6.47E-6 & mwasMeta$HetPVal > 0.05),]

phenInfo.sort<-phenInfo[order(phenInfo$SuperClass, phenInfo$Class, phenInfo$SubClass), ]
phenInfo.sort <- phenInfo.sort[(phenInfo.sort$UnifiedName %in% unique(mwasMeta.sig$Phenotype)), ]
phenInfo.sort$phenOrder<-c(1:nrow(phenInfo.sort))

speInfo.sort<-speInfo[order(speInfo$Kingdom ,speInfo$Phylum, speInfo$Class, speInfo$Order, speInfo$Family, speInfo$Genus, speInfo$Species),]
speInfo.sort$CumLen<-cumsum(speInfo.sort$`Reference Genome Size`) %>% .[-length(.)] %>% c(0,.)

gwasSig<-mwasMeta.sig[,c("Phenotype", "SpeciesId", "BP", "SNP", "P-value")]
gwasSig$BP<-mwasMeta.sig$SNP %>% str_replace_all(".*_", "")
colnames(gwasSig)<-c("Phenotype", "Chr", "Pos", "SNP", "P")
gwasSig$Chr<-as.numeric(gwasSig$Chr)
gwasSig$Pos<-as.numeric(gwasSig$Pos)
gwasSig$neglog10_P<- -log10(gwasSig$P)


gwasSig$phenOrder <- match(gwasSig$Phenotype, phenInfo.sort$UnifiedName) %>% phenInfo.sort$phenOrder[.]
gwasSig$Class<- match(gwasSig$Phenotype, phenInfo.sort$UnifiedName) %>% phenInfo.sort$Class[.]
gwasSig$Order<- match(gwasSig$Chr, speInfo.sort$`Species ID`) %>% speInfo.sort$Order[.]
gwasSig$CumLen<- match(gwasSig$Chr, speInfo.sort$`Species ID`) %>% speInfo.sort$CumLen[.]
gwasSig$GeneralPos<-gwasSig$CumLen+gwasSig$Pos
gwasSig$ImpactLevel<-match(gwasSig$SNP, snpAnno$V1) %>% snpAnno$ImpactLevel[.]
gwasSig$SigLevel <- "SpeWideSig"
gwasSig$SigLevel[gwasSig$P < metaWideP] <- "MetaWideSig"
gwasSig$SigLevel[gwasSig$P < studyWideP] <- "StudyWideSig"

gwasSig<-na.omit(gwasSig)

gwasSig$Color[gwasSig$Class=="Anthropology"]<-"#257CB2"
gwasSig$Color[gwasSig$Class=="CurrentExposure"]<-"#E93E3F"
gwasSig$Color[gwasSig$Class=="Diet"]<-"#4F9F3B"
gwasSig$Color[gwasSig$Class=="EarlyLifeExposure"]<-"#FE8B15"
gwasSig$Color[gwasSig$Class=="Stool"]<-"#10aec2"
gwasSig$Color[gwasSig$Class=="GeographySocioeconomics"]<-"#A6CEE3"
gwasSig$Color[gwasSig$Class=="HealthDisease"]<-"#F68181"
gwasSig$Color[gwasSig$Class=="MedicalMeasurement"]<-"#A5D981"
gwasSig$Color[gwasSig$Class=="MedicationUse"]<-"#F7F599"
gwasSig$Color[gwasSig$Class=="Metabolite"]<-"#73489F"

gwasSig$Color<-factor(gwasSig$Color)

Cairo::CairoPDF("09.metaAnalysisOverview/EMMAX.mwasMeta.summary.heatmap.pdf", width = 10, height = 10)
ggplot(gwasSig, aes(phenOrder, GeneralPos))+
  annotate("rect", xmin = 0,     xmax = 3.5,   ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#E93E3F", alpha.f = 0.2) )+
  annotate("rect", xmin = 3.5,   xmax = 14.5,  ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#4F9F3B", alpha.f = 0.2) )+
  annotate("rect", xmin = 14.5,  xmax = 15.5,  ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#FE8B15", alpha.f = 0.2) )+
  annotate("rect", xmin = 15.5,  xmax = 18.5,  ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#F7F599", alpha.f = 0.2) )+
  annotate("rect", xmin = 18.5,  xmax = 26.5,  ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#257CB2", alpha.f = 0.2) )+
  annotate("rect", xmin = 26.5,  xmax = 30.5,   ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#F68181", alpha.f = 0.2) )+
  annotate("rect", xmin = 30.5,  xmax = 61.5,  ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#A5D981", alpha.f = 0.2) )+
  annotate("rect", xmin = 61.5,  xmax = 103.5,   ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#73489F", alpha.f = 0.2) )+
  annotate("rect", xmin = 103.5, xmax = 105.5,   ymin = 0, ymax = sum(speInfo$`Reference Genome Size`),fill = adjustcolor("#10aec2", alpha.f = 0.2) )+
  geom_vline(xintercept = 2,  linetype = "solid", color = "grey80")+ # CurrentSmoker
  geom_vline(xintercept = 4,  linetype = "solid", color = "grey80")+ # DailyCalorieIntake
  geom_vline(xintercept = 6,  linetype = "solid", color = "grey80")+ # AlcoholFreqPerMonth
  geom_vline(xintercept = 7,  linetype = "solid", color = "grey80")+ # AlcoholGlassPerDay
  geom_vline(xintercept = 19, linetype = "solid", color = "grey80")+ # Age
  geom_vline(xintercept = 20, linetype = "solid", color = "grey80")+ # BMI
  geom_vline(xintercept = 22, linetype = "solid", color = "grey80")+ # Weight
  geom_vline(xintercept = 23, linetype = "solid", color = "grey80")+ # HipCircumference
  geom_vline(xintercept = 24, linetype = "solid", color = "grey80")+ # WaistCircumference
  geom_vline(xintercept = 25, linetype = "solid", color = "grey80")+ # WHR
  geom_vline(xintercept = 28, linetype = "solid", color = "grey80")+ # Hypertension
  geom_vline(xintercept = 40, linetype = "solid", color = "grey80")+ # BloodCreatinine
  geom_vline(xintercept = 41, linetype = "solid", color = "grey80")+ # Glucose
  geom_vline(xintercept = 44, linetype = "solid", color = "grey80")+ # HDL
  geom_vline(xintercept = 46, linetype = "solid", color = "grey80")+ # TG
  geom_vline(xintercept = 51, linetype = "solid", color = "grey80")+ # DBP
  geom_vline(xintercept = 52, linetype = "solid", color = "grey80")+ # SBP
  geom_vline(xintercept = 104, linetype = "solid", color = "grey80")+ # BristolType
  geom_vline(xintercept = 105, linetype = "solid", color = "grey80")+ # BristolFreq


#  annotate("rect", xmin = 0, xmax = 105.5,   ymin = 860231055, ymax = 863303105,fill = adjustcolor("grey50", alpha.f = 0.2) )+
#  annotate("rect", xmin = 0, xmax = 105.5,   ymin = 461070431, ymax = 463926731,fill = adjustcolor("grey50", alpha.f = 0.2) )+
#  annotate("rect", xmin = 0, xmax = 105.5,   ymin = 639355358, ymax = 642971118,fill = adjustcolor("grey50", alpha.f = 0.2) )+
  geom_point(aes(color = Class), alpha = 0.7)+
  xlab("Phenotypes")+
  ylab("SNPs")+
  scale_color_manual(breaks = phenClassOrder,
                     values = phenClassColor)+
  #theme_pubclean()+
  theme(legend.position = "none",
        axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank())
dev.off()
```

### 3.6 SNV overlap between phenotypes

```{r 3.6}
## Overlap SNVs between body shape-related phenotypes
phen.list <- c("Weight", "WaistCircumference", "BMI", "WHR", "HipCircumference")
phen.snp <- NULL
for (i in phen.list) {
#  i <- "BMI"
  phen.snp[[i]] <- unique(mwasMeta.cl.pt.full.filtered$SNP[mwasMeta.cl.pt.full.filtered$Phenotype==i])
}

phen.snp.df <- phen.snp %>%
  enframe(name = "Phenotype", value = "SNP") %>%
  unnest(SNP) %>%
  mutate(Presence = 1) %>%
  pivot_wider(names_from = Phenotype, values_from = Presence, values_fill = list(Presence = 0))
phen.snp.df$ImpactLevel <- phen.snp.df$SNP %>% match(snpAnno$V1) %>% snpAnno$ImpactLevel[.]

pdf("09.metaAnalysisOverview/bodyShape.assocSNV.inter.upset.pdf", height = 4, width = 8)
ComplexUpset::upset(phen.snp.df, phen.list,
                    mode='inclusive_intersection',
                    width_ratio=0.2,
                    base_annotations=list(
                      'Intersection size'=ComplexUpset::intersection_size(
                        counts=T,mode='inclusive_intersection',
                        mapping=aes(fill=ImpactLevel)
                      )+
                        scale_fill_manual(values = rev(impactLevelCol))+
                        theme_classic())
                    )
dev.off()

phen.df <- full_phen[,phen.list]
phen.df.cor.r <- cor(phen.df, use = "pairwise.complete.obs", method = "spearman")
phen.df.cor.p <- corrplot::cor.mtest(phen.df, conf.level = 0.95)

pdf("09.metaAnalysisOverview/bodyShape.cor.heatmap.pdf", height = 4.5, width = 6.5)
corrplot::corrplot(phen.df.cor.r, p.mat = phen.df.cor.p$p,
                   type = 'lower', insig='blank', addCoef.col ='black',number.cex = 0.8,
                   tl.srt = 45, tl.col = 'black', cl.pos = 'r', diag = T)
dev.off()
```

## 4 Functional proporty

### 4.1 Enrichment of high/moderate impact SNVs

```{r 4.1}
mwasMeta.spe.phen.enrich <- data.frame(Phenotype = mwasMeta.cl.pt.summ.filt$Var1, 
                                       Species = mwasMeta.cl.pt.summ.filt$Var2,
                                       AssocSnpN = mwasMeta.cl.pt.summ.filt$TotalN,
                                       AssocLowSnpN = mwasMeta.cl.pt.summ.filt$Low,
                                       AssocHighModerateSnpN = mwasMeta.cl.pt.summ.filt$High+mwasMeta.cl.pt.summ.filt$Moderate,
                                       AssocHighModerateProp = mwasMeta.cl.pt.summ.filt$HighModerateProp,
                                       TotalSnpN = mwasMeta.cl.pt.summ.filt$Var2 %>% match(snpAnnoSumm.filt$Var1) %>% snpAnnoSumm.filt$TotalN[.],
                                       TotalLowSnpN = mwasMeta.cl.pt.summ.filt$Var2 %>% match(snpAnnoSumm.filt$Var1) %>% snpAnnoSumm.filt$Low[.],
                                       TotalHighModerateSnpN = snpAnnoSumm.filt$High[match(mwasMeta.cl.pt.summ.filt$Var2, snpAnnoSumm.filt$Var1)] + 
                                                               snpAnnoSumm.filt$Moderate[match(mwasMeta.cl.pt.summ.filt$Var2, snpAnnoSumm.filt$Var1)],
                                       TotalLowProp = mwasMeta.cl.pt.summ.filt$Var2 %>% match(snpAnnoSumm.filt$Var1) %>% snpAnnoSumm.filt$LowProp[.],
                                       TotalHighModerateProp = mwasMeta.cl.pt.summ.filt$Var2 %>% match(snpAnnoSumm.filt$Var1) %>% snpAnnoSumm.filt$HighModerateProp[.])

mwasMeta.spe.phen.enrich$Chisq.P <- NA
mwasMeta.spe.phen.enrich$Log2Ratio <- NA

for (i in 1:nrow(mwasMeta.spe.phen.enrich)) {
#  i<-1
  chisq_i <- chisq.test(x = c(mwasMeta.spe.phen.enrich$AssocLowSnpN[i], mwasMeta.spe.phen.enrich$AssocHighModerateSnpN[i]),
                        p = c(mwasMeta.spe.phen.enrich$TotalLowProp[i], mwasMeta.spe.phen.enrich$TotalHighModerateProp[i]))
  mwasMeta.spe.phen.enrich$Chisq.P[i] <- chisq_i$p.value
  mwasMeta.spe.phen.enrich$Log2Ratio[i]   <- log2(mwasMeta.spe.phen.enrich$AssocHighModerateProp[i]/mwasMeta.spe.phen.enrich$TotalHighModerateProp[i])
}

mwasMeta.spe.phen.enrich$Chisq.FDR <- p.adjust(mwasMeta.spe.phen.enrich$Chisq.P, method = "fdr")
mwasMeta.spe.phen.enrich <- mwasMeta.spe.phen.enrich %>% dplyr::filter(!is.infinite(Log2Ratio))
mwasMeta.spe.phen.enrich$Sig <- "No"
mwasMeta.spe.phen.enrich$Sig[mwasMeta.spe.phen.enrich$Chisq.FDR < 0.05] <- "Yes"
write.table(mwasMeta.spe.phen.enrich,"09.metaAnalysisOverview/mwasMeta.spe.phen.enrich.tsv", sep = "\t", row.names = F, col.names = T, quote = F)

mwasMeta.spe.phen.enrich.label <- paste(mwasMeta.spe.phen.enrich$Phenotype,mwasMeta.spe.phen.enrich$Species,sep = "-")
mwasMeta.spe.phen.enrich.label[mwasMeta.spe.phen.enrich$Chisq.FDR>0.05] <- ""

pdf("09.metaAnalysisOverview/mwasMeta.spe.phen.enrich.volcano.pdf",width = 5.5, height = 4.5)
ggplot(mwasMeta.spe.phen.enrich,aes(Log2Ratio, -log10(Chisq.P), label = mwasMeta.spe.phen.enrich.label,color = Sig))+
  geom_point(aes(size = AssocSnpN), alpha = 0.5)+
  geom_text_repel(color = "black", box.padding = 0.5)+
  xlab('log2(Observed/Expected)')+
  ylab('-log10(P)')+
  scale_color_manual(name   = NULL,
                     breaks = c("Yes", "No"),
                     labels = c("Significant    ", "Insignificant"),
                     values = c("#ff4040","#4f94cd"))+
  scale_shape_discrete(name = NULL)+
  theme_bw()+
  theme(legend.position = "right",
        legend.key = element_rect(fill = NA))
dev.off()
```

### 4.2 Enzyme classes

```{r 4.2}
mwasMeta.ec <- mwasMeta.cl.pt.full.filtered %>% 
  dplyr::select(EC_number, ImpactLevel) %>%
  dplyr::filter(!is.na(EC_number)) %>%
  table() %>%
  as.data.frame() %>%
  dplyr::filter(Freq > 0)

mwasMeta.ec <- data.frame(ImpactLevel = as.character(mwasMeta.ec$ImpactLevel),
                          EC_level_2  = as.character(sub("\\.[^.]+\\.[^.]*$", "", mwasMeta.ec$EC_number)),
                          EC_level_3  = as.character(sub("\\.[^.]*$", "", mwasMeta.ec$EC_number)),
                          EC_level_4  = as.character(mwasMeta.ec$EC_number),
                          Count       = mwasMeta.ec$Freq)
mwasMeta.ec$ImpactLevel[mwasMeta.ec$ImpactLevel == "High"] <- "High/Moderate"
mwasMeta.ec$ImpactLevel[mwasMeta.ec$ImpactLevel == "Moderate"] <- "High/Moderate"

ec4 <- unique(mwasMeta.ec$EC_level_4)
for (i in 1:length(ec4)) {
#  i<-1
  if(sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_4 == ec4[i]]) <= 20){
    mwasMeta.ec$EC_level_4[mwasMeta.ec$EC_level_4 == ec4[i]] <- "Level 4 Others"
  }else{
    mwasMeta.ec$EC_level_4[mwasMeta.ec$EC_level_4 == ec4[i]] <- paste(ec4[i], " (", sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_4 == ec4[i]]), ")",sep = "")
  }
}

ec3 <- unique(mwasMeta.ec$EC_level_3)
for (i in 1:length(ec3)) {
#  i<-1
  if(sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_3 == ec3[i]]) <= 20){
    mwasMeta.ec$EC_level_3[mwasMeta.ec$EC_level_3 == ec3[i]] <- "Level 3 Others"
  }else{
    mwasMeta.ec$EC_level_3[mwasMeta.ec$EC_level_3 == ec3[i]] <- paste(ec3[i], " (", sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_3 == ec3[i]]), ")",sep = "")
  }
}

ec2 <- unique(mwasMeta.ec$EC_level_2)
for (i in 1:length(ec2)) {
  #  i<-1
  if(sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_2 == ec2[i]]) <= 20){
    mwasMeta.ec$EC_level_2[mwasMeta.ec$EC_level_2 == ec2[i]] <- "Level 2 Others"
  }else{
    mwasMeta.ec$EC_level_2[mwasMeta.ec$EC_level_2 == ec2[i]] <- paste(ec2[i], " (", sum(mwasMeta.ec$Count[mwasMeta.ec$EC_level_2 == ec2[i]]), ")",sep = "")
  }
}

sankey_colors<-c("#E64B35", "#3C5488")

pdf("09.metaAnalysisOverview/EC_sankey_plot.pdf",width = 4,height = 3.5)
ggplot(mwasMeta.ec,aes(axis1 = mwasMeta.ec$ImpactLevel, 
                       axis2 = mwasMeta.ec$EC_level_2, 
                       axis3 = mwasMeta.ec$EC_level_3,
                       axis4 = mwasMeta.ec$EC_level_4,
                       y     = mwasMeta.ec$Count))+
  scale_x_discrete(limits = c("Impact", "EC level 2", "EC level 3", "EC level 4")) +
  geom_alluvium(aes(fill = mwasMeta.ec$ImpactLevel),alpha = .6)+
  geom_stratum(alpha = 0,color = adjustcolor( "white", alpha.f = 0.95),size=0.5)+
  geom_text(stat = "stratum",cex=2,aes(label = after_stat(stratum))) +
  scale_color_manual(values=sankey_colors)+
  scale_fill_manual(values=sankey_colors)+
  theme_minimal()+
  theme(legend.position="none",
        axis.text = element_text(size = 10),
        axis.title = element_blank(),
        axis.text.y = element_blank(),
        panel.grid=element_blank())

dev.off()
```

### 4.3 Enzyme-phenotype

```{r 4.3}
mwasMeta.ec.phen <- mwasMeta.cl.pt.full.filtered %>% 
  dplyr::select(EC_number, ImpactLevel, Phenotype) %>%
  dplyr::filter(!is.na(EC_number)) %>%
  table() %>%
  as.data.frame() %>%
  dplyr::filter(Freq > 0)

mwasMeta.ec.phen.df <- pivot_wider(mwasMeta.ec.phen, names_from = "ImpactLevel", values_from = "Freq")
mwasMeta.ec.phen.df[is.na(mwasMeta.ec.phen.df)] <- 0
mwasMeta.ec.phen.df$TotalSnpN <- mwasMeta.ec.phen.df$Low + mwasMeta.ec.phen.df$Moderate + mwasMeta.ec.phen.df$High
```

